src/wp-includes/l10n.php

```diff
@@ -788,22 +788,81 @@ function load_textdomain( $domain, $mofile, $locale = null ) {
 		$locale = determine_locale();
 	}
 
-	$mo = new MO();
-	if ( ! $mo->import_from_file( $mofile ) ) {
-		$wp_textdomain_registry->set( $domain, $locale, false );
+	// Ensures the correct locale is set as the current one, in case it was filtered.
+	WP_I18n_Translation_Controller::instance()->set_locale( $locale );
 
-		return false;
+	/**
+	 * Filters the preferred file format for translation files.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string $preferred_format Preferred file format. Possible values: 'php', 'mo'. Default: 'php'.
+	 */
+	$preferred_format = apply_filters( 'translation_file_format', 'php' );
+	if ( ! in_array( $preferred_format, array( 'php', 'mo' ), true ) ) {
+		$preferred_format = 'php';
 	}
 
-	if ( isset( $l10n[ $domain ] ) ) {
-		$mo->merge_with( $l10n[ $domain ] );
+	$mofile_preferred = str_replace( '.mo', ".mo.$preferred_format", $mofile );
+
+	if ( 'mo' !== $preferred_format || str_ends_with( $mofile, $preferred_format ) ) {
+
+		/** This filter is documented in wp-includes/l10n.php */
+		$mofile_preferred = apply_filters( 'load_textdomain_mofile', $mofile_preferred, $domain );
+
+		/**
+		 * Filters the file path for loading translations for the given text domain.
+		 *
+		 * The file could be an MO or PHP file.
+		 *
+		 * @since 6.5.0
+		 *
+		 * @param string $file   Path to the translation file to load.
+		 * @param string $domain The text domain.
+		 */
+		$mofile_preferred = apply_filters( 'load_translation_file', $mofile_preferred, $domain );
```

Comment:
Similar to the other new filter, what's the rationale / use-case for introducing this filter?

Reply:
Let's combine the threads

Reply:
@swissspidy It's another use-case though, I don't think you address this on the other filter thread. Or maybe I missed it?

Reply:
So this is just like the load_textdomain_mofile filter, but the file being passed could also be a PHP file. For back compat I‘m not reusing the same filter.

Reply:
Makes sense, let's keep this one.